Modules/IO/ImageBase/src/itkImageIOFactory.cxx

```diff
@@ -19,53 +19,103 @@
 #include "itkImageIOFactory.h"
 
 #include <mutex>
-
+#include <vector>
 
 namespace itk
 {
 
 namespace
 {
-std::mutex createImageIOLock;
+std::vector<ImageIOBase::Pointer>
+GetAllImageIOInstances()
+{
+  static std::mutex           createImageIOLock;
+  std::lock_guard<std::mutex> mutexHolder(createImageIOLock);
+
+  const auto                        allInstances = ObjectFactoryBase::CreateAllInstance("itkImageIOBase");
+  std::vector<ImageIOBase::Pointer> result;
+  result.reserve(allInstances.size());
+  for (auto & possibleImageIO : allInstances)
+  {
+    auto * io = dynamic_cast<ImageIOBase *>(possibleImageIO.GetPointer());
+    if (io)
+    {
+      result.emplace_back(io);
+    }
+    else
+    {
+      std::cerr << "Error ImageIO factory did not return an ImageIOBase: " << possibleImageIO->GetNameOfClass() << '\n';
+    }
+  }
+  return result;
 }
+} // namespace
 
 ImageIOBase::Pointer
 ImageIOFactory::CreateImageIO(const char * path, FileModeType mode)
 {
-  std::list<ImageIOBase::Pointer> possibleImageIO;
-
-  std::lock_guard<std::mutex> mutexHolder(createImageIOLock);
+  if (mode != FileModeType::ReadMode && mode != FileModeType::WriteMode)
+  {
+    return {};
+  }
 
-  for (auto & allobject : ObjectFactoryBase::CreateAllInstance("itkImageIOBase"))
+  auto       imageIOInstances = GetAllImageIOInstances();
+  const bool ignoreCase = true;
+  // check instances that support file extension
+  for (auto & imageIO : imageIOInstances)
   {
-    auto * io = dynamic_cast<ImageIOBase *>(allobject.GetPointer());
-    if (io)
+    if (mode == FileModeType::ReadMode)
     {
-      possibleImageIO.emplace_back(io);
+      if (imageIO->HasSupportedReadExtension(path, ignoreCase))
+      {
+        if (imageIO->CanReadFile(path))
+        {
+          return imageIO;
+        }
+        else
+        {
+          imageIO = nullptr; // don't check it again
+        }
+      }
     }
     else
     {
-      std::cerr << "Error ImageIO factory did not return an ImageIOBase: " << allobject->GetNameOfClass() << std::endl;
+      if (imageIO->HasSupportedWriteExtension(path, ignoreCase))
+      {
+        if (imageIO->CanWriteFile(path))
+        {
+          return imageIO;
+        }
+        else
+        {
+          imageIO = nullptr;
+        }
+      }
     }
   }
-  for (auto & k : possibleImageIO)
+
+  for (auto & imageIO : imageIOInstances)
   {
+    if (!imageIO)
+    {
+      continue;
+    }
     if (mode == FileModeType::ReadMode)
     {
-      if (k->CanReadFile(path))
+      if (imageIO->CanReadFile(path))
       {
-        return k;
+        return imageIO;
       }
     }
-    else if (mode == FileModeType::WriteMode)
+    else
     {
-      if (k->CanWriteFile(path))
+      if (imageIO->CanWriteFile(path))
       {
-        return k;
+        return imageIO;
       }
     }
   }
-  return nullptr;
+  return {};
```

Comment:
I haven't seen `return {};` syntax before, so it is a bit confusing. I think that `return nullptr;` is clearer.